application/commandline_interface.py

```diff
@@ -131,64 +135,77 @@ def display_main_menu(self): # TESTED AND WORKING
     def change_password(self): # TESTED AND WORKING
         """Changes user password"""
         # Ask user for reauthentication
-        username, password = self.request_login_details()
+        password = self.request_password()
         # Authenticate user credentials
-        if self.login_service.login(username, password):
-            # Ask user for new password
-            new_password = getpass.getpass("New Password: ", stream=None)
-            # Ask user to confirm new password
-            new_password_confirmation = getpass.getpass("Confirm New Password: ", stream=None)
-            # Check if new password matches confirmation
-            if new_password == new_password_confirmation:
-                # Sanitise password
-                new_password = self.sanitisation_service.filter_special_characters(new_password)
-                # Check if new password is the same as the old password
-                if new_password == password:
-                    print("New password cannot be the same as the old password.\n")
-                else:
-                    # Ask user for phrase
-                    phrase = input("Enter your phrase: ")
-                    # Authenticate phrase
-                    if self.login_service.authenticate_phrase(phrase):
-                        # Change password
-                        self.login_service.change_password(new_password)
-                        # Create logger to log password change
-                        return True
+        if self.login_service.check_password(password):
+            self.login_service.display_password_requirements()
+            while True:
+                # Ask user for new password
+                new_password = getpass.getpass("New Password: ", stream=None)
+                # Check that new password is within constraints
+                if not self.sanitisation_service.validate_password(new_password):
+                    continue
+                # Ask user to confirm new password
+                new_password_confirmation = getpass.getpass("Confirm New Password: ", stream=None)
+                # Check if new password matches confirmation
+                if new_password == new_password_confirmation:
+                    # Sanitise password
+                    new_password = self.sanitisation_service.filter_special_characters(new_password)

```

Comment:
after filtering for special characters, if the new password is different to the password before being filtered, should the user be asked to create a different password?

Reply:
I think that's a good idea!

How about, to prevent this, the "sanitisation_service.validate_password(new_password)" method ensures that the password passed to it does not include any special characters. If it does, then the method won't allow the user to progress to the next step until they have entered a password that fulfills the requirements (i.e. doesn't include special characters)?

Reply:
Yeah, I think that is a good solution. Otherwise, I would worry that if the filtered password was different to what the user was expecting, they might not be able to log in again

Reply:
That makes sense! Thanks for pointing that out!